logbook-jdkserver/src/main/java/org/zalando/logbook/jdkserver/Request.java

```diff
@@ -0,0 +1,231 @@
+            return exchange.getRequestBody();
+        }
+
+        default byte[] getBody() {
+            return new byte[0];
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Unbuffered implements State {
+
+        @Override
+        public State with() {
+            return new Offering();
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Offering implements State {
+
+        @Override
+        public State without() {
+            return new Unbuffered();
+        }
+
+        @Override
+        public State buffer(HttpExchange exchange) throws IOException {
+            return new Buffering(toByteArray(exchange.getRequestBody()));
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static abstract class Streaming implements State {
+
+        @Getter(PROTECTED)
+        private final ByteArrayInputStream stream;
+
+        @Override
+        public InputStream getInputStream(final HttpExchange exchange) throws IOException {
+            return stream;
+        }
+
+    }
+
+    private static final class Buffering extends Streaming {
+
+        private final byte[] body;
+
+        Buffering(final byte[] body) {
+            this(body, new ByteArrayInputStream(body));
+        }
+
+        Buffering(final byte[] body, final ByteArrayInputStream stream) {
+            super(stream);
+            this.body = body;
+        }
+
+        @Override
+        public State without() {
+            return new Ignoring(body);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return body;
+        }
+
+    }
+
+    private static final class Ignoring extends Streaming {
+
+        private final byte[] body;
+
+        Ignoring(final byte[] body) {
+            super(new ByteArrayInputStream(body));
+            this.body = body;
+        }
+
+        @Override
+        public State with() {
+            return new Buffering(body, getStream());
+        }
+
+    }
+
+    public Request(HttpExchange httpExchange) {
+        this.httpExchange = httpExchange;
+        this.state = new AtomicReference<>(new Unbuffered());
+    }
+
+    @Override
+    public Origin getOrigin() {
+        return Origin.REMOTE;
+    }
+
+    @Override
+    public HttpHeaders getHeaders() {
+        HttpHeaders headers = HttpHeaders.empty();
+        for (Map.Entry<String, List<String>> e : httpExchange.getRequestHeaders().entrySet()) {
+            headers = headers.update(e.getKey(), e.getValue());
+        }
+        return headers;
+    }
+
+    @Override
+    public String getRemote() {
+        return httpExchange.getRemoteAddress().getHostString();
+    }
+
+    @Override
+    public String getMethod() {
+        return httpExchange.getRequestMethod();
+    }
+
+    @Override
+    public String getScheme() {
+        String scheme = httpExchange.getRequestURI().getScheme();
+        if (scheme == null) {
+            scheme = "http";
+        }
+        return scheme;
```

Comment:
I think it's safe to fall back to empty string here as in other modules.
```suggestion
        return notNull(httpExchange.getRequestURI().getScheme());
```